File: Top200/234-IsPalindrome.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func isPalindrome(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return true
	}

	dummy := &ListNode{Val: -1}
	slow := dummy
	fast := dummy
	dummy.Next = head

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	reverseHead := reverseList(slow.Next)
	slow.Next = nil

	for reverseHead != nil {
		if head.Val != reverseHead.Val {
			return false
		}

		head = head.Next
		reverseHead = reverseHead.Next
	}

	return true
}

func reverseList(head *ListNode) *ListNode {
	var prev *ListNode = nil
	curr := head

	for curr != nil {
		temp := curr.Next
		curr.Next = prev
		prev = curr
		curr = temp
	}

	return prev
}

File: Top200/234-IsPalindrome.java
package Top200;

class IsPalindrome {
    class ListNode {
        int val;
        ListNode next;
        
        ListNode() {}
        
        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public boolean isPalindrome(ListNode head) {
        ListNode dummy = new ListNode(-1);
        ListNode slow = dummy;
        ListNode fast = dummy;
        dummy.next = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode reverseHead = slow.next;
        slow.next = null;

        reverseHead = this.reverseList(reverseHead);
        ListNode curr = head;

        while (reverseHead != null) {
            if (curr.val != reverseHead.val) {
                return false;
            }
            
            curr = curr.next;
            reverseHead = reverseHead.next;
        }

        return true;
    }

    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode prev = null;
        ListNode curr = head;

        while (curr != null) {
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }

        return prev;
    }
}
